src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,655 @@
+                }
+            // falls through
+            case Diagnostics.Rest_parameter_0_implicitly_has_an_any_type.code:
+                return getCodeActionForParameters(<ParameterDeclaration>token.parent);
+
+            // Get Accessor declarations
+            case Diagnostics.Property_0_implicitly_has_type_any_because_its_get_accessor_lacks_a_return_type_annotation.code:
+            case Diagnostics._0_which_lacks_return_type_annotation_implicitly_has_an_1_return_type.code:
+                return isGetAccessor(containingFunction) ? getCodeActionForGetAccessor(containingFunction) : undefined;
+
+            // Set Accessor declarations
+            case Diagnostics.Property_0_implicitly_has_type_any_because_its_set_accessor_lacks_a_parameter_type_annotation.code:
+                return isSetAccessor(containingFunction) ? getCodeActionForSetAccessor(containingFunction) : undefined;
+
+            // Property declarations
+            case Diagnostics.Member_0_implicitly_has_an_1_type.code:
+                return getCodeActionForVariableDeclaration(<VariableDeclaration>token.parent);
+        }
+
+        return undefined;
+
+        function getCodeActionForVariableDeclaration(declaration: VariableDeclaration | PropertyDeclaration | PropertySignature) {
+            if (!isIdentifier(declaration.name)) {
+                return undefined;
+            }
+
+            const type = inferTypeForVariableFromUsage(declaration.name);
+            const typeString = type && typeToString(type, declaration);
+
+            if (!typeString) {
+                return undefined;
+            }
+
+            return createCodeActions(declaration.name.getText(), declaration.name.getEnd(), `: ${typeString}`);
+        }
+
+        function getCodeActionForVariableUsage(token: Identifier) {
+            const symbol = checker.getSymbolAtLocation(token);
+            return symbol && symbol.valueDeclaration && getCodeActionForVariableDeclaration(<VariableDeclaration>symbol.valueDeclaration);
+        }
+
+        function isApplicableFunctionForInference(declaration: FunctionLike): declaration is MethodDeclaration | FunctionDeclaration | ConstructorDeclaration {
+            switch (declaration.kind) {
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.Constructor:
+                    return true;
+                case SyntaxKind.FunctionExpression:
+                    return !!(declaration as FunctionExpression).name;
+            }
+            return false;
+        }
+
+        function getCodeActionForParameters(parameterDeclaration: ParameterDeclaration): CodeAction[] {
+            if (!isIdentifier(parameterDeclaration.name) || !isApplicableFunctionForInference(containingFunction)) {
+                return undefined;
+            }
+
+            const types = inferTypeForParametersFromUsage(containingFunction) ||

```

Comment:
If we do the function body first we are guranteed to manifacture something that the makes the funciton happy, but it usually will be too general. giving usage precedence allows us to get more specific types.. e.g.:

```ts
function length(a) { return a.length; }

length([1,2,3]);
```

inferring from function body will give `a` a type `{ length: any}`. which is accurate, but rarely what the user would write manually.

inferring from the call site gives us `number[]` which is more like it..

Reply:
Perhaps you should infer `<T extends { length }>(a: T) => ...` instead. In combination with return type inference, this should yield `<T extends { length }>(a: T) => T["length"]`, which will work correctly when applied to arrays (or indeed anything with a `length` property).

Reply:
I tried that.. the types get unwieldy very fast.. and you have this structural type that has `length`, `push`, `splice`, with an element that is has `kind`, instead of just `Node[]`. the call sites provided the more user-identifiable name.. though the first one is more correct.